### PR TITLE
as_strided batching rule

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47227 Batched gradient support for view+inplace operations
* #47226 Batching rule for Tensor.new_empty_strided
* #47225 Implement Tensor.new_empty_strided(sizes, strides, *, dtype, device, requires_grad)
* **#47224 as_strided batching rule**
* #47223 Batched grad for advanced indexing (index)
* #47189 Fix sum batching rule, add simple clone batching rule
* #47188 Batching rules for complex view functions

This PR adds a batching rule for as_strided. `as_strided` is a really weird
operation and I hope that users don't use it very much.

Motivation
----------
The motivation for adding a batching rule for as_strided is for
batched gradient computation.

AsStridedBackward appears in PyTorch when handling view+in-place
operations and calls `as_strided`. AsStridedBackward calls as_strided on
a fresh tensor with storage_offset equal to 0. We would like to be able
to vmap through the backward graph of view+in-place operations to
for batched gradient computation, especially because internally we have
a number of functions that are implemented as a view+in-place.

Alternatives
------------
If we think that as_strided is too crazy to have a batching rule, we
could either:
- have a flag that controls the autograd view+in-place
behavior
- require that the input tensor's storage offset must be equal to 0
to make it easier to reason about.

I think the batching rule makes sense, so I didn't pursue the
alternatives.

The batching rule
-----------------
```
y = vmap(lambda x: x.as_strided(sizes, strides, offset))(xs)
```
The result of the above should be "equivalent" to:
- Assume that each x has storage offset equal to xs.storage_offset()
(call that S).
- Calling as_strided with (sizes, sizes, offset + x[i].storage_offset() - S) on each x.

More concretely,
this returns a view on `xs`, such that each y[i] has:
- sizes: `sizes`
- strides: `strides`
- storage_offset: offset + i * x.stride(batch_dim)

Why the behavior can be weird
-----------------------------
The behavior of the batching rule may be different from actually running
as_strided in a for-loop because `as_strided` takes in `offset` as a
"absolute offset". As an example, consider

```
>>> x = torch.tensor([0., 1., 2., 3., 4.])
>>> z = [x[i].as_strided([1], [1], 1) for i in range(5)]
```
Each z[i] is actually the same view on x (z[i] == torch.tensor([0.]))!
However, we consider the above for-loop comprehension to be a user error:
a user should have written the following if they wanted to use as_strided
in a per-sample way:
```
>>> z = [x[i].as_strided([1], [1], 1 + x[i].storage_offset()) for i in range(4)]
```

Test Plan
---------
- Added some tests that compare vmap+as_strided to vmap+(the equivalent operator)

### CODE REVIEW

**Comment by albanD** [comment]
In your example above, you should have `z[i] == torch.tensor([1.])` (and not `0.`) right? As the offset you give is `1`?

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 1]
```diff
@@ -344,6 +344,135 @@ Tensor view_batching_rule(const Tensor& self, IntArrayRef size) {
...
+  auto num_batch_dims = physical_view.numBatchDims();
+  auto physical_sizes = physical_view.getPhysicalShape(sizes);
+
+  // physical_strides = physical tensor's batch strides + (logical) strides
+  at::VmapDimVector physical_strides;
+  physical_strides.reserve(num_batch_dims + strides.size());
+  physical_strides.insert(
+      physical_strides.end(),
+      physical_view.tensor().strides().begin(),
+      physical_view.tensor().strides().begin() + num_batch_dims);
```
Comment (albanD) [inline aten/src/ATen/BatchingRegistrations.cpp:384]: There is no ways the two successive calls to `physical_view.tensor().strides()` would return different objects right?

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 2]
```diff
@@ -344,6 +344,135 @@ Tensor view_batching_rule(const Tensor& self, IntArrayRef size) {
...
+// - offset: offset
+//
+// xs.as_strided([B] + sizes, [S] + strides, offset)[i] has:
+// - sizes: sizes
+// - strides: strides
+// - offset: offset + S * i
+// These memory locations are exactly the same as what we got for [[A]],
+// so the xs.as_strided([B] + sizes, [S] + strides, offset) is valid.
+//
+// [[B]] Hand-wavy proof of Claim 1:
```
Comment (albanD) [inline aten/src/ATen/BatchingRegistrations.cpp:449]: nit `[[B]]` here is a bit confusing with the `[B]` size used for the sizes. Maybe numbers or `*` would be better?

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 3]
```diff
@@ -344,6 +344,135 @@ Tensor view_batching_rule(const Tensor& self, IntArrayRef size) {
...
+// Part of our definition of being valid is that xs[i].as_strided(...)
+// must return a tensor that only uses memory indexable by xs[i].
+// This means that (sizes, strides, offset + xs[i].offset() - xs.offset()) satisfies:
+//    offset + xs[i].offset() - xs.offset() + 1 + \sum_j (sizes[j] - 1) * strides[j]
+//    <= xs[i].offset() + 1 + \sum_j (xs[i].size(j) - 1) * xs[i].stride(j)
+// (the largest-index memory location of xs[i].as_strided(...) must be \leq
```
Comment (albanD) [inline aten/src/ATen/BatchingRegistrations.cpp:455]: nit: You can mention that under the assumption we're making, the lower bound (lowest indexed memory) is trivially within the storage.

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 4]
```diff
@@ -344,6 +344,135 @@ Tensor view_batching_rule(const Tensor& self, IntArrayRef size) {
...
+// If zi = xs[i].as_strided(sizes, strides, offset + xs[i].offset() - xs.offset())
+// is valid for all i, then it turns out that
+// xs.as_strided(physical_sizes, physical_strides, offset) always succeeds and
+// creates a tensor y such that each y[i] refers to the same memory as zi.
+//
+// Let's say we have xs[i].as_strided(sizes, strides, offset + xs[i].offset() - xs.offset()).
+// Furthermore, let's say that as a part of being "valid" this as_strided call
+// does not return a result that can index memory not indexable by xs[i].
+//
+// Assume that there's only one batch dim and it is at the front of the
```
Comment (albanD) [inline aten/src/ATen/BatchingRegistrations.cpp:410]: Is that assumption "without loss of generatlity"? Or does it change when the batch dimension is not at the front or when there is more than one?

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 5]
```diff
@@ -344,6 +344,135 @@ Tensor view_batching_rule(const Tensor& self, IntArrayRef size) {
...
+
+// NOTE: [When will the as_strided batching rule fail?]
+// If zi = xs[i].as_strided(sizes, strides, offset + xs[i].offset() - xs.offset())
+// is valid for all i, then it turns out that
+// xs.as_strided(physical_sizes, physical_strides, offset) always succeeds and
+// creates a tensor y such that each y[i] refers to the same memory as zi.
+//
+// Let's say we have xs[i].as_strided(sizes, strides, offset + xs[i].offset() - xs.offset()).
+// Furthermore, let's say that as a part of being "valid" this as_strided call
+// does not return a result that can index memory not indexable by xs[i].
```
Comment (albanD) [inline aten/src/ATen/BatchingRegistrations.cpp:408]: Should we add some sanity checks in the batching rule to reduce this risk? Like making sure that the original as_strided wasn't indexing outside of the input?

**Review by albanD — COMMENTED** [review]
I am not sure what will be the general structure for the vmap doc. Should we plan to add a note there about `as_strided`?

File: aten/src/ATen/BatchingRegistrations.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/BatchingRegistrations.cpp:384]: > There is no ways the two successive calls to physical_view.tensor().strides() would return different objects right?

They do return different IntArrayRef objects. Now there is a question of how the iterators work. I will change this to use the same strides() to avoid potential bugs

File: aten/src/ATen/BatchingRegistrations.cpp (hunk 4 above)
Comment (zou3519) [reply aten/src/ATen/BatchingRegistrations.cpp:410]: This is without loss of generality.
- If the batch dim is not at the front, we can move it to the front and then go through the contents here (we do not assume a particular stride for xs or the batch dim here).
- For multiple batch dims, this works out similarly. It's just a bit annoying to write because if there are K batch dimensions then we need K indices I0 I1... Ik, and `S*i` becomes `offset + \sum_{i=1}^{k} S_i * I_k`

**Comment by zou3519** [comment]
I needed to move this PR in the stack. ghstack doesn't like it when PRs are removed, so it turned into a brand-new PR: https://github.com/pytorch/pytorch/pull/47364
